src/geometry/utility/geometry.cc

```diff
@@ -272,6 +274,61 @@ LineString3d ComputeCenterline3d(const LineString3d& left, const LineString3d& r
   return LineString3d{centerline};
 }
 
+Vector3 InterpolatedPointAtP(const LineString3d& line_string, double p) {
+  // Implementation inspired on:
+  // https://github.com/fzi-forschungszentrum-informatik/Lanelet2/blob/master/lanelet2_core/include/lanelet2_core/geometry/impl/LineString.h#L618
+  static constexpr double kEpsilon{1e-12};
+  if (p < 0) return line_string.first();
+  if (p >= line_string.length()) return line_string.last();
+
+  const auto line_string_points_length = GetBoundPointsAtP(line_string, p);
+  const double partial_length{(*line_string_points_length.first - *line_string_points_length.second).norm()};
+  const double remaining_distance = p - line_string_points_length.length;
+  if (remaining_distance < kEpsilon) {
+    return *line_string_points_length.first;
+  }
+  return *line_string_points_length.first +
+         remaining_distance / partial_length * (*line_string_points_length.second - *line_string_points_length.first);
+}
+
+double GetSlopeAtP(const LineString3d& line_string, double p) {
+  const BoundPointsResult bound_points = GetBoundPointsAtP(line_string, p);
+  const double dist{(To2D(*bound_points.second) - To2D(*bound_points.first)).norm()};
+  const double delta_z{bound_points.second->z() - bound_points.first->z()};
+  MALIPUT_THROW_UNLESS(!(dist == 0. && delta_z == 0.));
+  return delta_z / dist;
```

Comment:
However, we could be facing an inf as a return statement...
Let's log at least a warning when `dist == 0` so we know that two consecutive points are equal.

Reply:
Also, we could consider making sure in the LineString constructor that it does not happen that so we can also THROW here when you find it.

Reply:
> However, we could be facing an inf as a return statement...

An inf would be valid as slope.(Not a physically possible one tho) You could have two points that are equal in x and y but with a different z.

> Also, we could consider making sure in the LineString constructor that it does not happen that so we can also THROW here when you find it.

We definitely could add that requirement in the line string's constructor. I feel like we freely could handle it either way without adding that constraint and that verification (what if we have thousands of points per line string?).

When calling `GetBoundPointsAtP` it will return both points that confine the passed `p`. When p is equal to a point that defines the line string it will return `point_at_p` and `point_at_p - 1`. And the order matters here, it starts from the start with this algorithm.
Having said that even tho if we have two equal consecutive points in the line string we won't face this issue here because those two points won't be taken as the pair to be evaluated.

Reply:
Sounds good.